File: flaskr/utils/date_utils.py
import requests as rq
import datetime
import time

URL_ID = 'https://hacker-news.firebaseio.com/v0/item/{}.json?print=pretty'
URL_MAXID = 'https://hacker-news.firebaseio.com/v0/maxitem.json?print=pretty'

def date2ts(year: int, month: int, day: int) -> int:
    """
    return timestamp corresponding to date given by (year, month, day)
    """
    date = datetime.date(year=year, month=month, day=day)
    timetpl = date.timetuple()
    return int(time.mktime(timetpl))

def id2ts(item_id: int) -> int:
    """
    get timestamp of hn item with specified id
    """
    res = rq.get(URL_ID.format(item_id))
    while not res.ok or not res.json().get('time'):
        item_id += 1
        res = rq.get(URL_ID.format(item_id))
    return res.json()['time']

def check(year: int, month: int, day: int, item_id: int) -> bool:
    """
    confirm that hn item with specified id was posted on date
    specified by (year, month, day)
    """
    ts = id2ts(item_id)
    d = datetime.datetime.fromtimestamp(ts)    
    print(f'first id on {year}/{month}/{day}: {item_id} ({d.year}/{d.month}/{d.day})')            

    return year == d.year and month == d.month and day == d.day      

def get_first_id_on_day(year: int, month: int, day: int) -> int:
    """
    returns first hn item id (story or comment) on specified date;
    date specified as as (year, month, day) tuple if ints (1-based indexing), 
    e.g. (2021, 1, 1) is 1st January 2021;
    impossible dates (e.g., 31st February 2021) or future date (e.g., 1st January 3021)
    will raise errors
    """
    # convert date to timestamp
    target_ts = date2ts(year, month, day)
    max_id = rq.get(URL_MAXID).json()
    max_ts = id2ts(max_id)

    if target_ts > max_ts:
        raise ValueError('Specified date is out of range')
    
    # use binary search to find first id higher than timestamp
    lo, hi = 1, max_id    
    while lo <= hi:
        mi = (lo + hi) // 2        
        ts = id2ts(mi)
        if target_ts > ts:
            lo = mi + 1
        else:
            hi = mi - 1

    #print(check(year, month, day, hi+1))          
    return hi + 1

File: flaskr/static/js/date_utils.js
function adjustDateNodeDisplay(node, date) {
    /*
    Print *actual* date in node's inner's html:
    e.g., if 2021/feb/31 was selected
    it would be corrected to 2021/3/3
    */
    const pattern = /[0-9]+\/[0-9]+\/[0-9]+/;
    const r = pattern.exec(node.innerHTML);
    if (r) {
        node.innerHTML = 
            node.innerHTML.replace(
                r[0], `${date.getFullYear()}/${date.getMonth()+1}/${date.getDate()}`
            );
    } else {
        node.innerHTML =     
            node.innerHTML.split(':')[0] + ': ' +
            `${date.getFullYear()}/${date.getMonth()+1}/${date.getDate()} ` +
            node.innerHTML.split(':')[1];
    }
}
 

function getDateFromDateNode(node) {
    /*
    reads year/month/day from node's children 
    if children with id's containing year/month/day are present
    and their values are numerical,
    then returns Date object;
    otherwise throws an error 
    */
    let date = {
        'year': undefined, 
        'month': undefined,
        'day': undefined
    };

    for (let child of node.children) {
        for (let name in date) {
            if (child.id && child.id.includes(name)) {
                date[name] = child.value;
            }
        }
    }

    if (Object.keys(date)
        .map(k => isFinite(date[k]))
        .reduce((a,b) => a && b)
    ) {
        const [y, m, d] = [
            date.year, date.month, date.day
        ].map(val => parseInt(val));
        
        const dateObj = new Date(y, m-1, d);
        adjustDateNodeDisplay(node, dateObj);

        return dateObj;
    } else {
        throw new Error('date is not specified');
    }
}

function getTimestampFromDateNode(node) {
    /*
    attempts to read year/month/day from the date node 
    and adjusts node inner html to reflect the *actual* date
    (e.g., if 2021/feb/31 was selected it would be corrected to 2021/3/3);
    on success returns timestamp in seconds(!);
    on fail throws `date is not specified` error and returns undefined
    */
    const date = getDateFromDateNode(node);
    return isFinite(date.valueOf()) ? parseInt(date.valueOf() / 1000) : undefined;
}

function date2ts(year, month, day) {
    /* returns timestamp (in seconds) corresponding to specified date */
    const dateObj = new Date(year, month-1, day);
    return Math.floor(dateObj.getTime() / 1000);
}

async function id2ts(itemId) {
    /* returns timestamp (in seconds) of hn item with specified id */
    const url = `https://hacker-news.firebaseio.com/v0/item/${itemId}.json`;
    const res = await fetch(url)
        .then(res => res.json())
        .catch(err => console.log(err));
    
    if (res && res['time']) {
        return res['time'];
    } else {
        return await id2ts(itemId - 1)
        .then(res => res.json());
    }
}

async function binSearchTs(lo, hi, targetTs) {
    // use binsearch to find first id higher than timestamp
    let mi, ts;
    while (lo <= hi) {
        mi = Math.floor((lo + hi) / 2);
        
        await id2ts(mi).then(res => {ts = res}); // await!!!
        
        if (targetTs > ts) {
            lo = mi + 1;
        } else {
            hi = mi - 1;
        }
    }
    return hi + 1;
}

async function fetchFirstIdOnDay(year, month, day) {
    /*
    returns first hn item id (story or comment) on specified date;
    date specified as as (year, month, day) tuple if ints (1-based indexing),
    e.g. (2021, 1, 1) is 1st January 2021;
    impossible dates (e.g., 31st February 2021) will be resolved
    (e.g. 31st February 2021 -> 3rd March 2021),
    future dates (e.g., 1st January 3021) will raise errors

    use:
        const firstId = await fetchFirstIdOnDay(...);
    */
    URL_MAXID = 'https://hacker-news.firebaseio.com/v0/maxitem.json';

    const targetTs = date2ts(year, month, day);

    let maxId, maxTs;
    return await fetch(URL_MAXID)
    .then(res => res.json())
    .then(res => maxId = res)
    .then(async (res) => {        
        await id2ts(maxId).then(res => {maxTs = res}); // await!!!
        if (targetTs > maxTs) {
            //throw new Error('Specified date is out of range');
            console.log(
                `Specified date (${new Date(targetTs * 1000).toISOString()}) is out of range, ` +
                `using current date ${new Date(maxTs * 1000).toISOString()} instead.`
            );
            return maxId;
        } else {
            return await binSearchTs(1, maxId, targetTs);
        }
    })
    .catch(err => console.log(err));
}
